### PR TITLE
PyTorchShim: add support for mixed-precision

### PR DESCRIPTION
**Pull request by danieldk**
This change adds support for mixed-precision training and prediction to
the PyTorch shim. Mixed-precision support consists of two parts:

- Autocasting of the forward pass. This changes whitelisted ops to run
  in half precision rather than single precision.
- Gradient scaling. Gradients underflow quickly in half-precision.
  Gradient scaling increases the magnitude of the gradients during
  backprop to avoid underflow.

A separate class, PyTorchGradScaler is introduced to perform gradient
scaling and store the associated state.

---

I hope that I correctly implemented multi-GPU training, but I couldn't test it with my current single GPU workstation.

### CODE REVIEW

File: thinc/layers/pytorchwrapper.py [hunk 1]
```diff
@@ -34,6 +34,8 @@ def PyTorchWrapper(
     pytorch_model,
     convert_inputs: Optional[Callable] = None,
     convert_outputs: Optional[Callable] = None,
+    mixed_precision: bool = False,
+    grad_scaler: Optional[PyTorchGradScaler] = None,
```
Comment (svlandeg) [inline thinc/layers/pytorchwrapper.py:38]: We need to increment the versioning here to `PyTorchWrapper.v2` to make sure it's clear that these two parameters are only supported in the newer version and from a specific Thinc release. Both definitions can be here in the same file though.

File: thinc/shims/pytorch_grad_scaler.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                device_tensors, found_inf_device, inv_scale_device
+            )
+
+            self._found_inf += found_inf_device.to(self._found_inf.device)
+
+        return bool(self._found_inf != 0)
+
+    def update(self):
+        """
+        Update the the scale factor and clear information about infinities.
```
Comment (svlandeg) [inline thinc/shims/pytorch_grad_scaler.py:116]:
```suggestion
        Update the scale factor and clear information about infinities.
```

**Review by svlandeg — COMMENTED** [review]
Some first quick comments:

File: thinc/layers/pytorchwrapper.py (hunk 1 above)
Comment (danieldk) [reply thinc/layers/pytorchwrapper.py:78]: Ah, right, should be fixed now.

**Comment by honnibal** [comment]
It's hard to love the mechanism here. 🤔  It's just sort of disappointing to need this helper class.

I don't really have a better suggestion though, and it doesn't affect the rest of the API, it's specific to the PyTorchShim. So I'm okay with merging this as-is. I'd definitely try to find some other way of doing things if it were on the main `Model` class though!

**Comment by svlandeg** [comment]
We can have it as an ondocumented beta feature for now, allowing us to still revisit the implementation details at some point if we want. So I'll go ahead and merge!
